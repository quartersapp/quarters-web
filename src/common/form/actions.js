import { createAction } from 'redux-actions'

import { CHANGE_FORM_VALUE, REGISTER_FORM, DEREGISTER_FORM } from './types'

export const changeFormValue = createAction(
  CHANGE_FORM_VALUE,
  (formName, field, value) => {
    return { form: formName, field, value }
  }
)

export const registerForm = createAction(
  REGISTER_FORM,
  (formName, initialValues = {}) => {
    return { form: formName, initialValues }
  }
)

export const deregisterForm = createAction(
  DEREGISTER_FORM,
  (formName) => {
    return { form: formName }
  }
)
