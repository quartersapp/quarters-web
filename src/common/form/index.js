export { default as createForm } from './create-form'
export { default as createFormField } from './create-form-field'
export { default as reducer } from './reducer'
export {
  createFormValuesSelector,
  createFormIsValidSelector
} from './selectors'
