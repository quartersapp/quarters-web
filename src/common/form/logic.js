import PropTypes from 'prop-types'
import { kea } from 'kea'

export default (options = {}) => {
  const { form } = options

  if (!form) throw new Error('form name must be set')

  return kea({
    path: () => ['form', form],

    actions: () => ({
      changeFormValue: (form, field, value) => ({ form, field, value }),
      registerForm: form => ({ form }),
      deregisterForm: form => ({ form })
    }),

    reducers: ({ actions }) => ({
      values: [{ actions }, PropTypes.object.isRequired, {
        [actions.changeFormValues]: (state, { payload: { field, value } }) => {
          return Object.assign({}, state, { [field]: value })
        }
      }]
    })
  })
}
