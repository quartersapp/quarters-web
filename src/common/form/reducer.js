import { createReducer } from 'redux-create-reducer'
import { combineReducers } from 'redux'
import { static as Immutable } from 'seamless-immutable'
import {
  CHANGE_FORM_VALUE,
  REGISTER_FORM,
  DEREGISTER_FORM
} from './types'

const formReducer = combineReducers({
  initialValues: createReducer({}, {}),
  values: createReducer({}, {
    [CHANGE_FORM_VALUE] (state, { payload: { field, value } }) {
      return Immutable.set(state, field, value)
    }
  }),
  numRegisteredForms: createReducer(0, {
    [REGISTER_FORM]: state => state + 1,
    [DEREGISTER_FORM]: state => state - 1
  })
})

const initialState = Immutable({})

const formsReducer = (optionsPerForm = {}) => {
  return (state = initialState, action) => {
    const { type, payload } = action

    if (type === REGISTER_FORM) {
      const { form } = payload

      if (!state[form] && optionsPerForm[form] && optionsPerForm[form].initialValues) {
        state = Immutable.setIn(state, [form, 'initialValues'], optionsPerForm[form].initialValues)
      }

      return Immutable.set(state, form, formReducer(state[form], action))
    } else if (type === DEREGISTER_FORM) {
      const { form } = payload
      const nextFormState = formReducer(state[form], action)
      if (nextFormState.numRegisteredForms > 0) {
        return Immutable.set(state, form, nextFormState)
      } else {
        return Immutable.without(state, form)
      }
    } else if (type === CHANGE_FORM_VALUE && state[payload.form]) {
      return Immutable.set(state, payload.form, formReducer(state[payload.form], action))
    } else { // proxy all remaining actions to each formReducer
      return state
    }
  }
}

export default formsReducer
