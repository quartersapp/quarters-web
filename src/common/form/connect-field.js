import React, { Component } from 'react'
import { connect } from 'react-redux'
import PropTypes from 'prop-types'

import { changeFormValue } from './actions'

export default WrappedComponent => {
  class FormField extends Component {
    constructor (props, context) {
      super(props, context)

      if (!context.form) {
        throw new Error(
          'Field must be rendered inside of a Form'
        )
      }
    }

    getFormName () {
      return this.context.form
    }

    handleChange = e => {
      e && e.preventDefault()
      this.props.changeFormValue(this.getFormName(), this.props.name, e.target.value)
    }

    render () {
      const form = this.getFormName()
      const field = this.props.name
      const value = this.props.formState[form].values[field]

      return (
        <WrappedComponent {...this.props} onChange={this.handleChange} value={value} />
      )
    }
  }

  FormField.propTypes = {
    name: PropTypes.string.isRequired,
    formState: PropTypes.object.isRequired
  }

  FormField.contextTypes = {
    form: PropTypes.string
  }

  return connect(
    state => ({ formState: state.form }),
    { changeFormValue }
  )(FormField)
}
