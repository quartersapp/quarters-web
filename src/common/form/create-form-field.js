import React, { Component } from 'react'
import { connect } from 'react-redux'
import PropTypes from 'prop-types'

import { changeFormValue } from './actions'

const createFormField = () => WrappedComponent => {
  class FormField extends Component {
    getFormName () {
      return this.context.form
    }

    handleChange = e => {
      e && e.preventDefault()
      this.props.changeFormValue(this.getFormName(), this.props.name, e.target.value)
    }

    render () {
      const form = this.getFormName()
      const field = this.props.name
      const value = this.props.formState[form].values[field]

      return (
        <WrappedComponent {...this.props} onChange={this.handleChange} value={value} />
      )
    }
  }

  FormField.propTypes = {
    name: PropTypes.string.isRequired,
    formState: PropTypes.object.isRequired
  }

  FormField.contextTypes = {
    form: PropTypes.string
  }

  return connect(
    state => ({ formState: state.form }),
    { changeFormValue }
  )(FormField)
}

export default createFormField
