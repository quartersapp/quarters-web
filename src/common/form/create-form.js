import React, { Component } from 'react'
import { connect } from 'react-redux'
import { createStructuredSelector } from 'reselect'
import PropTypes from 'prop-types'
import { registerForm, deregisterForm } from './actions'
import { createFormValuesSelector } from './selectors'

export const createForm = name => {
  class Form extends Component {
    componentWillMount () {
      this.props.registerForm(name)
    }

    componentWillUnmount () {
      this.props.deregisterForm(name)
    }

    /**
     * Set the form name to the context so that field
     * components have access to the form name
     */
    getChildContext () {
      return { form: name }
    }

    handleSubmit = e => {
      e.preventDefault()
      this.props.onSubmit && this.props.onSubmit(this.props.values)
    }

    render () {
      return (
        <form onSubmit={this.handleSubmit}>
          {this.props.children}
        </form>
      )
    }
  }

  Form.childContextTypes = {
    form: PropTypes.string
  }

  Form.propTypes = {
    onSubmit: PropTypes.func,
    children: PropTypes.node
  }

  return Form
}

export default name => connect(
  createStructuredSelector({ values: createFormValuesSelector(name) }),
  { registerForm, deregisterForm }
)(createForm(name))
