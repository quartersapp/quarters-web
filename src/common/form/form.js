import React, { PureComponent } from 'react'
import { connect } from 'react-redux'
import { createSelector, createStructuredSelector } from 'reselect'
import Immutable from 'seamless-immutable'
import PropTypes from 'prop-types'
import {
  compose,
  setPropTypes,
  withContext,
  defaultProps,
  setDisplayName
} from 'recompose'
import { initializeForm, destroyForm, renameForm } from './actions'

export const formValuesSelector = createSelector(
  (state, props) => props.name,
  state => state.form,
  (formName, formState) => Immutable.static.getIn(formState, [formName, 'values'], {})
)

export const Form = compose(
  setDisplayName('Form'),
  defaultProps({ destroyOnUnmount: true }),
  setPropTypes({
    children: PropTypes.node,
    name: PropTypes.string.isRequired,
    onSubmit: PropTypes.func,
    /* If set to true, will dispatch action to destroy form state on unmount */
    destroyOnUnmount: PropTypes.bool.isRequired
  }),
  withContext(
    { form: PropTypes.string },
    props => ({ form: props.name })
  )
)(class Form extends PureComponent {
  componentWillMount () {
    this.props.initializeForm(this.props.name)
  }

  componentWillReceiveProps (nextProps) {
    if (nextProps.name !== this.props.name) {
      this.props.renameForm(this.props.name, nextProps.name)
    }
  }

  componentWillUnmount () {
    if (this.props.destroyOnUnmount) {
      this.props.destroyForm(this.props.name)
    }
  }

  handleSubmit = e => {
    e.preventDefault()
    if (this.props.onSubmit) {
      this.props.onSubmit(this.props.values)
    }
  }

  render () {
    return (
      <form onSubmit={this.handleSubmit}>{this.props.children}</form>
    )
  }
})

export default connect(
  createStructuredSelector({ values: formValuesSelector }),
  { initializeForm, renameForm, destroyForm }
)(Form)
