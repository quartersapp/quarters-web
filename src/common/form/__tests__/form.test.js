/* eslint-env jest */

import React from 'react'
import PropTypes from 'prop-types'
import { mount } from 'enzyme'
import renderer from 'react-test-renderer'
import Immutable from 'seamless-immutable'
import { Form, formValuesSelector } from '../form'

const noop = () => {}
const FORM_NAME = 'testForm'

describe('Form', () => {
  it('matches snapshot', () => {
    const tree = renderer.create(
      <Form name={FORM_NAME} registerForm={noop}>
        Form goes here
      </Form>
    )
    expect(tree.toJSON()).toMatchSnapshot()
  })

  it('sets the form name to the context', () => {
    /* Couldn't get context testing with enzyme working, so this will have to do */
    expect(Form.childContextTypes).toEqual({ form: PropTypes.string })
    expect(new Form({ name: FORM_NAME }).getChildContext()).toEqual({ form: FORM_NAME })
  })

  it('dispatches register and deregister actions on mount/unmount', () => {
    const registerForm = jest.fn()
    const deregisterForm = jest.fn()

    const wrapper = mount(
      <Form name={FORM_NAME} registerForm={registerForm} deregisterForm={deregisterForm} />
    )
    expect(registerForm).toBeCalled()
    expect(deregisterForm).not.toBeCalled()

    wrapper.unmount()
    expect(deregisterForm).toBeCalled()
  })

  it('dispatches a move action the form if the name prop changes', () => {
    const moveRegisteredForm = jest.fn()
    const NEW_FORM_NAME = 'newForm'

    const wrapper = mount(
      <Form name={FORM_NAME} registerForm={noop} moveRegisteredForm={moveRegisteredForm} />
    )
    expect(moveRegisteredForm).not.toBeCalled()
    wrapper.setProps({ name: NEW_FORM_NAME })
    expect(moveRegisteredForm).toBeCalledWith(FORM_NAME, NEW_FORM_NAME)
  })

  it('calls props.handleSubmit with the form values onSubmit', () => {
    const handleSubmit = jest.fn()
    const preventDefault = jest.fn()
    const formValues = {
      email: 'test@example.com',
      password: '12345'
    }

    const wrapper = mount(
      <Form name={FORM_NAME} registerForm={noop} onSubmit={handleSubmit} values={formValues} />
    )

    wrapper.find('form').simulate('submit', { preventDefault })
    expect(handleSubmit).toBeCalledWith(formValues)
    expect(preventDefault).toBeCalled()
  })
})

describe('formValuesSelector', () => {
  let state

  beforeEach(() => {
    state = Immutable.static({
      form: {
        [FORM_NAME]: {
          values: {
            email: 'test@example.com',
            password: '12345'
          }
        }
      }
    })
  })

  it('selects the form values for the current form name', () => {
    const props = { name: FORM_NAME }

    expect(formValuesSelector(state, props)).toEqual({
      email: 'test@example.com',
      password: '12345'
    })
  })
})
