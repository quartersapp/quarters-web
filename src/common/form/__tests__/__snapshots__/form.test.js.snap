// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Form matches snapshot 1`] = `
<form
  onSubmit={[Function]}
>
  Form goes here
</form>
`;
